Check if the discriminator value is `null` before converting it and setting into `rowData` during hydration

fullbl opened the pull request:
fixes #6303

Ocramius commented:
Needs tests

On 24 Feb 2017 7:30 p.m., "fullbl" <notifications@github.com> wrote:

> fixes #6303 <https://github.com/doctrine/doctrine2/issues/6303>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/doctrine/doctrine2/pull/6304
> Commit Summary
>
>    - check if value is null before converting it and setting into rowData
>
> File Changes
>
>    - *M* lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php
>    <https://github.com/doctrine/doctrine2/pull/6304/files#diff-0> (8)
>
> Patch Links:
>
>    - https://github.com/doctrine/doctrine2/pull/6304.patch
>    - https://github.com/doctrine/doctrine2/pull/6304.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/doctrine/doctrine2/pull/6304>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJakEit5cOVeTTHSLfvACICMkc5yCJBks5rfyHYgaJpZM4MLh1s>
> .
>

fullbl commented:
Tests added, but seems that my edit broke a lot of other tests... 

Checking for non empty values instead of isset seems to fix things, but this solution is dipendent on the default value of SimpleArray.
I really believe that value assignation should be made in 2 steps, one for checking null values and another one for assigning the default in case of null!

Ocramius commented on tests/Doctrine/Tests/Models/Issue5989/Issue5989Person.php:
```diff
@@ -10,6 +10,7 @@
  * @DiscriminatorMap({
  *      "person"    = "Issue5989Person",
  *      "manager"   = "Issue5989Manager",
+ *      "secretary"  = "Issue5989Secretary",
```
Please use a different entity and a new inheritance tree to test your specific scenario

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6303.php:
```diff
@@ -0,0 +1,68 @@
+        $secretary = new Issue5989Secretary();
+
+        $secretaryTags = 'single_tag';
+        $secretary->tags = $secretaryTags;
```
The fact that you use a different type for a property in the inheritance tree is quite confusing: is this intentional?

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6303.php:
```diff
@@ -0,0 +1,68 @@
+
+        static::assertEquals($tags[$people[0]->id], $people[0]->tags);
+        static::assertEquals($tags[$people[1]->id], $people[1]->tags);
+        static::assertEquals($tags[$people[2]->id], $people[2]->tags);
```
Which of these assertions would fail without your patch in place?

Ocramius commented on lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php:
```diff
@@ -290,10 +290,10 @@ protected function gatherRowData(array $data, array &$id, array &$nonemptyCompon
                     // in an inheritance hierarchy the same field could be defined several times.
                     // We overwrite this value so long we don't have a non-null value, that value we keep.
                     // Per definition it cannot be that a field is defined several times and has several values.
```
This particular definition is what is being challenged by your inheritance design. The field definition should be fetched per entity name, and `$fieldName` should be specific to the currently hydrated entity type instead of just the name of a field. Did you try step-debugging here? What is `$type`?

fullbl commented:
Hi, I will create new entities as soon as possible, I thought it would be good to reuse the old ones!

$type is StringType for string and SimpleArrayType for simple array, so is different based on the field.

The problem is exactly as you stated in 

> The fact that you use a different type for a property in the inheritance tree is quite confusing: is this intentional?

This is what is breaking stuff.


The failing assertion is the one for the string field. It's hydrated by the simple_array as an empty array and missing the real value.

To be honest, I'm not satisfied with my solution, I will think about it in next days!

fullbl committed:
tests with custom inheritance tree

fullbl commented:
Hi again, I fixed the tests to use their own inheritance, but I removed my fix, because I would like to add more tests for checking that setting some property to false, 0 or empty array should return correct values!

fullbl committed:
added tests on empty values

fullbl committed:
use dataMap also in first test function

fullbl committed:
check if array is set before assigning

fullbl commented:
Here we are, I added those tests and decided to redo the fix in another way, keeping only values from the correct table!

lcobucci commented on lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php:
```diff
@@ -368,13 +375,23 @@ protected function hydrateColumnInfo($key)
+                    'dqlAlias'            => $ownerMap,
+                    
                 ];
+                if( !empty($classMetadata->parentClasses && isset($this->_rsm->discriminatorColumns[$ownerMap]))){
```
Is this really the expression we're looking for? It looks like it should be:

```php
 if ( ! empty($classMetadata->parentClasses) && isset($this->_rsm->discriminatorColumns[$ownerMap])) {
 }
```

lcobucci reviewed (dismissed):
Can you also fix the coding style to match with rest of the project, please?

fullbl committed:
fix wrong pharenthesis

fullbl committed:
moved all test models into test class

fullbl committed:
removed trailing spaces

fullbl commented:
Done, I made it first into different files because I copypasted from another issue, maybe it was old!

Ocramius commented on lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php:
```diff
@@ -287,13 +287,20 @@ protected function gatherRowData(array $data, array &$id, array &$nonemptyCompon
+                    if(
+                        isset($cacheKeyInfo['discriminatorColumn']) && 
+                        isset($data[$cacheKeyInfo['discriminatorColumn']]) &&
+                        $data[$cacheKeyInfo['discriminatorColumn']] != $cacheKeyInfo['discriminatorValue']
```
This should be `!==` to avoid type juggling hell. Yes, that means that object discriminators will fail, that's an edge case that the ORM doesn't handle anyway.

fullbl replied:
it seems to me that this change is breaking another test (/code/tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheJoinTableInheritanceTest.php:179). Is it ok?

lcobucci replied:
@fullbl it shouldn't break it... let's keep the strict comparison and try to find the root cause.

Ocramius commented on lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php:
```diff
@@ -368,13 +375,23 @@ protected function hydrateColumnInfo($key)
+                    
                 ];
+                if( !empty($classMetadata->parentClasses) && isset($this->_rsm->discriminatorColumns[$ownerMap])){
+                    $returnArray += [
```
Consider making two `return` operations instead of using `+=`

Ocramius commented on tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6303Test.php:
```diff
@@ -0,0 +1,154 @@
+        $contractA->originalData = $contractAData;
+
+        $contractB = new DDC6303ContractB();
+        $contractBData = ['accepted', 'authorized'];
```
Could you add comments for the reader, specifying that these two entities map `originalData` differently?

Ocramius reviewed (approved):
Patch looks good to me! The rest are nitpicks and improvements that can be done on merge.

Commit:
clarified what's the problem in a comment

fullbl committed:
comparison on discriminator value with !==

fullbl committed:
Merge branch 'skip-hydrate-on-null' of github.com:fullbl/doctrine2 into skip-hydrate-on-null

Ocramius commented:
@fullbl I rebased this branch on top of `master` to start merging it, but there's a failure in the test suite:

```
1) Doctrine\Tests\ORM\Functional\SecondLevelCacheJoinTableInheritanceTest::testOneToManyRelationJoinTable
Failed asserting that null matches expected '0000-0000'.

/home/ocramius/Documents/doctrine/doctrine2/tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheJoinTableInheritanceTest.php:179

FAILURES!
Tests: 3185, Assertions: 11878, Failures: 1, Skipped: 46, Incomplete: 2.
```

Commit:
Update AbstractHydrator.php

revert strict comparison in hydration (fails on some dates)

fullbl commented:
I reverted the edit, it was introduced when you asked this: 

> This should be !== to avoid type juggling hell. Yes, that means that object discriminators will fail, that's an edge case that the ORM doesn't handle anyway.

Ocramius commented:
@fullbl interesting edge case indeed - wasn't expecting the change to break that...

Ocramius commented:
As per @lcobucci's [comment](https://github.com/doctrine/doctrine2/pull/6304#discussion_r124134568):

 > @fullbl it shouldn't break it... let's keep the strict comparison and try to find the root cause.

We should investigate further on that particular strict comparison, sorry :-(

fullbl commented:
So, after all I found the problem (I think):

The AttractionInfo has this DiscriminatorMap:

```
 * @DiscriminatorMap({
 *  1 = "AttractionContactInfo",
 *  2  = "AttractionLocationInfo",
 * })
```

Changing it into this, test goes green:

```
 * @DiscriminatorMap({
 *  "cInfo"  = "AttractionContactInfo",
 *  "lInfo"  = "AttractionLocationInfo",
 * })
```


So the problem should be that the ObjectHydrator is getting a string for the dtype column, while comparing with an integer. Setting the default DiscriminatorColumn type as integer into the AnnotationDriver doesn't solve it, because PDO will always return it as string.

What would be the problem in keeping loose comparison?

Ocramius commented:
I checked the issue locally, and it's because the discriminator is not converted to a PHP value. 

That said, trying to fix the issue leads to a massive can of worms that is being opened: let's keep the loose comparison there for now.

Ocramius commented:
Manually merged via doctrine/doctrine2@a30d8d85ea15239d58ea6376dd0ae6b750a15d3f

Thanks @fullbl! It took me a while to figure out, but basically, the bug depends on the discriminator map sorting.

A lot of weird factors involved, and indeed an extremely tricky bug to fix. Good work!

JanTvrdik commented:
Maybe I'm doing sth wrong but this change seems to broke the following multi-level STI schema when selecting with QueryBuilder `FROM C` because `$cacheKeyInfo['discriminatorValue']` is `NULL` and therefore `$data[$cacheKeyInfo['discriminatorColumn']] != $cacheKeyInfo['discriminatorValue']` returns in `FALSE`.

![image](https://user-images.githubusercontent.com/175109/29814280-18412aea-8cad-11e7-92df-854013b57475.png)
